### PR TITLE
417 reorder table column table

### LINKED ISSUE #417: Allow dragging of table view headers to reorder fields
From Mark Showalter:

You are looking at a table view. Each column has a header. The header is actually a pop-up menu. When you click on the header, you can select any other quantity that we can tabulate. When you release, the content of that column is updated based on the new selection.

In addition there's a widget at right that let's you add another column (or maybe another N columns all at once). They appear, and then you can select each header.

We would also need options to close columns and to sort by columns. I bet they could be integrated into the table headers as additional widgets.

We also need to be able to reorder columns. Maybe that could be done by dragging columns right a left, Otherwise, there could at least be a way to select one or more columns and click a button to move them right or left.

As a user, I find it a bit tedious to go to a separate page or an overlay when I want to choose columns. This UI would make the table view much more interactive and responsive.

### PR DESCRIPTION
**Pull request by dstopp**
- Fixes #417 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - If YES:
    - Database used: opus3_test_200729
    - All Django tests pass: Y/NA
    - FLAKE8 run on modified code: Y/NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: <still testing>
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): <still testing>

Description of changes:
 - added a dropdown toolbar to the table header that allows users to add either before or after the metadata element a new metadata field OR remove the selected metadata field.  

Known problems:
 - code for the reorder of the table headers is still in place but commented out

### CODE REVIEW

File: opus/application/static_media/js/utils.js [hunk 1]
```diff
@@ -154,3 +154,8 @@ $.fn.isOnScreen = function(scope, slop) {
 
     return (elementTop + offset <= bottom) && (elementTop >= top);
 };
+
+// more convienent this way...
```
Comment (rfrenchseti) [inline opus/application/static_media/js/utils.js]: Spell convenient

File: opus/application/apps/ui/templates/ui/base.html [hunk 2]
```diff
@@ -265,11 +265,18 @@ <h3 class="op-header-text"></h3>
...
+                    <a href="#" class="op-metadata-detail-remove" data-action="remove" title="Remove this metadata field"><i class="far fa-trash-alt"></i></a>
+                    <a href="#" class="op-metadata-detail-add" data-action="addAfter" title="Add a metadata field after this column"><i class="fas fa-plus pr-1"></i></a>
+                </div>
+            </div>
         </div><!-- /tabcontent -->
 
 
         <!-- dropdown menu for observations -->
-        <div class="dropdown-menu dropdown-menu-sm" id="op-obs-menu" data-toggle="dropdown" aria-labelledby="op-obs-menu" aria-haspopup="true" aria-expanded="false">
+        <div id="op-obs-menu" class="dropdown-menu dropdown-menu-sm"data-toggle="dropdown" aria-labelledby="op-obs-menu" aria-haspopup="true" aria-expanded="false">
```
Comment (rfrenchseti) [inline opus/application/apps/ui/templates/ui/base.html]: Missing space before data-toggle

File: opus/application/apps/ui/templates/ui/base.html [hunk 3]
```diff
@@ -281,8 +288,14 @@ <h2 class="dropdown-header"></h2>
             <a class="dropdown-item" data-action="downloadURL" href="#" download><i class="fas fa-download"></i>Download zipped URL archive&nbsp;</a>
         </div>
 
+        <!-- drodown menu for adding metadata fields -->
```
Comment (rfrenchseti) [inline opus/application/apps/ui/templates/ui/base.html]: Spell dropdown

File: opus/application/static_media/css/opus.css [hunk 4]
```diff
@@ -227,7 +227,7 @@ so that both "Recent Announcements" and "Message" can be vertically aligned. */
     cursor: grabbing;
 }
 
-.op-metadata-detail-add {
+.#galleryView .op-metadata-detail-add {
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Is there an extra "." before #galleryView?

File: opus/application/static_media/css/opus.css [hunk 5]
```diff
@@ -741,6 +756,22 @@ th.sticky-header {
     border-right-width: 15px;
 }
 
+/* Fine tune the sortable helper text to stay at bottom in table header.
+Note: the original text moves up because the helper's default position is
+absolute, and this will result its child element ui-resizable (position: relative)
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Bad grammar

File: opus/application/static_media/js/browse.js [hunk 6]
```diff
@@ -397,6 +402,57 @@ var o_browse = {
...
+            // only hide the edit metadata field tool bar if the add metadata menu is not showing
+            if (!$("#op-add-metadata-fields").hasClass("show")) {
+                o_browse.hideTableMetadataTools(e);
+            }
+        });
+
+        $("#op-edit-field-tool a").on("click", function(e) {
+            let action = $(this).data("action");
+            let slug = $("#op-edit-field-tool").data("slug");
+            switch(action) {
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Space after switch

File: opus/application/static_media/js/browse.js [hunk 7]
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
         });
     },
 
+    initDraggableColumn: function(tab) {
+        function moveColumn(table, from, to) {
+            let rows = $('tr', table);
+            let cols;
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Why is cols set outside of the loop?

File: opus/application/static_media/js/browse.js [hunk 8]
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
...
+                // maintain width of all headers while moving
+                $(e.currentTarget).find("th").each(function() {
+                    $(this).width($(this).width());
+                });
+
+                return ui.clone();
+            },
+            stop: function(e, ui) {
+                o_browse.isSortingHappening = false;
+                let columnOrder = $(this).sortable( "toArray" );
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Extra spaces

File: opus/application/static_media/js/browse.js [hunk 9]
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
...
+
+                // only bother if something actually changed...
+                if (!o_utils.areObjectsEqual(opus.prefs.cols, columnOrder)) {
+
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Extra blank line

File: opus/application/static_media/js/browse.js [hunk 10]
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
...
+    hideTableMetadataTools: function() {
+        $("#op-edit-field-tool").removeClass("show").hide();
+    },
+
+    onDoneUpdateFromTableMetadataDetails: function(e) {
+        o_utils.disableUserInteraction();
+
+        o_hash.updateURLFromCurrentHash(); // This makes the changes visible to the user
+        // passing in false indicates to not close the gallery view on loadData
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Passing in false to what?

File: opus/application/static_media/js/browse.js [hunk 11]
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
...
+        $("#op-edit-field-tool").data("slug", slug);
+    },
+
+    hideTableMetadataTools: function() {
+        $("#op-edit-field-tool").removeClass("show").hide();
+    },
+
+    onDoneUpdateFromTableMetadataDetails: function(e) {
+        o_utils.disableUserInteraction();
+
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Extra blank line

File: opus/application/static_media/js/browse.js [hunk 12]
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
...
+            }
         }
         return top;
     },
 
     showMetadataList: function(e) {
-        let contextMenu = "#op-add-metadata-fields";
-        let targetPosition = $(e.currentTarget).position();
-        let left = targetPosition.left;
+        let targetPosition = $(e.currentTarget).offset();
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js:2393]: Something in here seems to have changed. With a little work I can get the *slideshow* add-metadata dropdown to dropup when it shouldn't, and run a little off the top of the browser. If I have exactly the same screen on the production OPUS, it doesn't do it.

File: opus/application/static_media/js/browse.js [hunk 13]
```diff
@@ -2293,6 +2462,7 @@ var o_browse = {
     },
 
     checkForEmptyMetadataList: function() {
+        // if there is only one metadata field left, disable the trash in both the gallery view and table tools dropdown
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Gallery view? Do you mean slideshow?

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- I haven't found a way to reproduce it with 100% reliability, but if you add and delete columns a lot you eventually get to a state where the "add/delete" toolbar ends up on top of the header, instead of below. Once it's there, it never gets better. When this happens, it's also slightly offset from the header position - a few pixels to the right.

**Comment by dstopp** [comment]
All comments have been addressed except the one that causes the toolbar to get slightly offset.

**Review by rfrenchseti — APPROVED** [review]

**Pull request merged**